### ISSUE TITLE
Memory only exceeds 1G, so sad

### ISSUE THREAD

**Issue by shuaixn** [smthemex/ComfyUI_PhotoDoodle#13]
![Image](https://github.com/user-attachments/assets/254ff74d-4295-4535-9993-fa149dcc82da)

**Comment by smthemex** [comment]
use repo ，diffuser==0.32.2

**Comment by shuaixn** [comment]
> use repo ，diffuser==0.32.2使用 repo，diffuser==0.32.2

thanks，and what is repo? I tried to ask ai, such as grok and gpt, but they didn't understand and said it was an online service? Is that right? I couldn't find what it is on the Internet.

**Comment by smthemex** [comment]
download below all ：

![Image](https://github.com/user-attachments/assets/9eecc20e-bc8a-4b88-910f-e2ff3860a6c9)

**Comment by shuaixn** [comment]
> download below all ：
> 
> ![Image](https://github.com/user-attachments/assets/9eecc20e-bc8a-4b88-910f-e2ff3860a6c9)
Hi, thanks for your guidance on using the repo black-forest-labs/FLUX.1-dev. I have set up the model by manually downloading files (excluding ae.safetensors and flux1-dev.safetensors as suggested) into D:\AI\ComfyUI\models\checkpoints\FLUX1_DEV. The VRAM usage is under 10GB, and it takes about 30-42 seconds to generate an image (20 steps), which is great based on the log below. However, I have some concerns:
Unsatisfactory Effects: The generated images (prompt: "add a halo and wings for the cat by sksmagiceffects") don't show noticeable effects like halos or wings. They are either very subtle or absent. Did I miss a setup step? Should I adjust LoRA strength, CFG Scale, or include ae.safetensors for better VAE support?
Fixed Seed Behavior: After generating one image, clicking "Run" again produces the same image every time. Is this normal? How can I enable random generation? I couldn't find a randomize option in PhotoDoodle_Sampler.
by the way,When I start MMGP (using flux.dev.16g), the video memory usage does not change. I have mmgp in my environment. My diffuser == 0.32.2

Could you please advise on improving the generation quality, enabling random outputs, and confirming the correct model setup? Thanks!

![Image](https://github.com/user-attachments/assets/b5b1564e-05fc-4769-babc-a66c0f3eac15)

**Comment by smthemex** [comment]
By default, NF4 quantization is enabled, and the effect naturally declines. I will set a switch

**Comment by shuaixn** [comment]
> By default, NF4 quantization is enabled, and the effect naturally declines. I will set a switch

Hi, thank you for your response! I’ve tested the NF4 quantized black-forest-labs/FLUX.1-dev, but the generation quality is almost unusable. The effects (e.g., halo and wings on the cat with prompt "add a halo and wings for the cat by sksmagiceffects") are barely visible, far below my expectations.

Switching back to the local single-checkpoint model flux1-dev-fp8-16G.safetensors, the results are much better (clear halo and wings), but on my RTX 4090, it consumes 25GB of VRAM, exceeding my 24GB limit by about 1GB. Is it possible to optimize the memory usage to around 23GB to avoid overflow? For example, could we adjust model loading, reduce LoRA usage, or apply other optimizations?

I also tested the single-checkpoint flux-fp8 model (downloaded from Kijai/flux-fp8), loading CLIP and VAE separately, but the VRAM usage spiked to 34GB (exceeding by 10GB), which is much worse than flux1-dev-fp8-16G. 

Lastly, MMGP (mmgp 3.2.7) doesn’t seem to work. I’ve updated to the latest version via GitHub, but whether enabled or not, VRAM usage remains at 25GB,  Could you advise on how to properly enable MMGP? Are there other memory optimization suggestions?

**Comment by smthemex** [comment]
try fp8 repo

**Comment by shuaixn** [comment]
> try fp8 repo

I tried using the fp8 repo and the image was solid grey with no content, then I updated it and it looked like the switch was half done, right?

![Image](https://github.com/user-attachments/assets/734641e7-7fe1-4f28-abcb-fc4c17186899)

**Comment by smthemex** [comment]
> > try fp8 repo
> 
> I tried using the fp8 repo and the image was solid grey with no content, then I updated it and it looked like the switch was half done, right?
> 
> ![Image](https://github.com/user-attachments/assets/734641e7-7fe1-4f28-abcb-fc4c17186899)

更新插件后，请重新从菜单拉取节点。

**Comment by shuaixn** [comment]
> > > try fp8 repo
> > 
> > 
> > I tried using the fp8 repo and the image was solid grey with no content, then I updated it and it looked like the switch was half done, right?
> > ![Image](https://github.com/user-attachments/assets/734641e7-7fe1-4f28-abcb-fc4c17186899)
> 
> 更新插件后，请重新从菜单拉取节点。

Hi, thank you for your continued support! I’ve updated PhotoDoodle and noticed the new quantization option in PhotoDoodle_Loader. However, I’m encountering several issues with both the unquantized Flux repo (black-forest-labs/FLUX.1-dev) and the FP8 repo (nada-mah/flux-fp8) on my RTX 4090 (24GB VRAM). I’ve upgraded peft and diffusers to the latest versions, but the problems persist. Here’s a summary:

VRAM Overflow with quantization="none":
When I set quantization="none" for both the unquantized and FP8 repos, the VRAM usage exceeds my 24GB limit by about 1GB (around 25GB). 
Poor Effect Quality with NF4:
When I enable NF4 quantization, the effect changes are almost invisible (e.g., no visible halo or wings with prompt "add a halo and wings for the cat by sksmagiceffects"). I even tried increasing steps to 50, 80, and 150, but the effect remains unsatisfactory. 
Error with FP8 Quantization:
When I set quantization="fp8", I get the error 'FluxTransformer2DModel' object has no attribute 'peft_config'. I consulted an AI, which suggested modifying PhotoDoodle_node.py at line 119 to initialize peft_config, but I lack programming experience and couldn’t implement it. Could you provide a simpler solution or fix this in the next update?

**Comment by smthemex** [comment]
need update peft>=0.15.0
```
python -m pip install -U peft
```

**Comment by shuaixn** [comment]
> need update peft>=0.15.0
> 
> ```
> python -m pip install -U peft
> ```

Thank you so much for your guidance and support throughout this process! With your suggestions, I’ve finally managed to get PhotoDoodle working successfully.
Since all my issues (VRAM overflow, poor NF4 effects, and FP8 error) are resolved for now, I’ll close this issue. Thank you again for your amazing work on PhotoDoodle and for helping me through this!

![Image](https://github.com/user-attachments/assets/162a90e7-6b0c-434f-9cc4-85a34c5e8f48)

**Comment by shuaixn** [comment]
![Image](https://github.com/user-attachments/assets/d37bf76f-4f3b-4dc9-a202-963ecc9f884b)

> need update peft>=0.15.0
> 
> ```
> python -m pip install -U peft
> ```

Hi, I noticed during use that the random seed functionality doesn’t seem to be supported, so I added a random seed feature to the sampler. My Python skills are at a "hello world" level, so I made this modification with the help of Grok. I’m not sure if this change is the most optimal, but it works effectively for me. I’ve attached the modified PhotoDoodle_node.py file and a screenshot of my workflow.


```python
[# !/usr/bin/env python
# -*- coding: UTF-8 -*-
import os
import torch
import numpy as np
from safetensors.torch import load_file
from diffusers import AutoencoderKL,FluxTransformer2DModel
from diffusers import BitsAndBytesConfig as DiffusersBitsAndBytesConfig
from .node_utils import tensor2pil_list, load_images,cleanup
from .src.pipeline_pe_clone import FluxPipeline
from .src.pipeline_pe_clone_orgin import FluxPipeline as FluxPipeline_orgin

import folder_paths
from comfy.utils import ProgressBar
MAX_SEED = np.iinfo(np.int32).max
current_node_path = os.path.dirname(os.path.abspath(__file__))

device = torch.device(
    "cuda:0") if torch.cuda.is_available() else torch.device(
    "mps") if torch.backends.mps.is_available() else torch.device(
    "cpu")


class PhotoDoodle_Loader:
    def __init__(self):
        pass

    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "flux_unet": (["none"] + folder_paths.get_filename_list("diffusion_models"),),
                "vae": (["none"] + folder_paths.get_filename_list("vae"),),
                "pre_lora": (["none"] + [i for i in folder_paths.get_filename_list("loras") if "pre" in i],),
                "loras": (["none"] + folder_paths.get_filename_list("loras"),),
                "flux_repo":("STRING", {"default": "", "multiline": False}),
                "quantization":(["none","fp8","nf4"],),
                "use_mmgp":("BOOLEAN",{"default":False}),
                "profile_number":([1,2,3,4,5],)
            },
        }

    RETURN_TYPES = ("MODEL_PhotoDoodle",)
    RETURN_NAMES = ("model",)
    FUNCTION = "loader_main"
    CATEGORY = "PhotoDoodle"

    
    def loader_main(self,flux_unet,vae,pre_lora,loras,flux_repo,quantization,use_mmgp,profile_number):
       
        flux_repo_local=os.path.join(current_node_path, 'src/FLUX.1-dev')   
        print("***********Load model ***********")
        #load model
        
        if pre_lora != "none":
            pre_lora_path = folder_paths.get_full_path("loras", pre_lora)
        else:
            raise ValueError("No model selected")
    
        if loras != "none":
            lora_path = folder_paths.get_full_path("loras", loras)
        else:
            raise ValueError("No model selected")
        need_clip=False
        if  not flux_repo:
            if flux_unet != "none":
                flux_transformer_path = folder_paths.get_full_path("diffusion_models", flux_unet)
            else:
                raise ValueError("No model selected")
            if vae != "none":
                vae_path = folder_paths.get_full_path("vae", vae)
                vae_config=os.path.join(flux_repo_local, 'vae')
                ae = AutoencoderKL.from_single_file(vae_path,config=vae_config, torch_dtype=torch.bfloat16)
                config_file = os.path.join(flux_repo_local,"transformer/config.json")
                t_state_dict=load_file(flux_transformer_path)
                unet_config = FluxTransformer2DModel.load_config(config_file)
                transformer = FluxTransformer2DModel.from_config(unet_config).to(torch.bfloat16)
                transformer.load_state_dict(t_state_dict, strict=False)
                del t_state_dict
                cleanup()
                pipeline = FluxPipeline.from_pretrained(
                    flux_repo_local,
                    vae=ae,
                    transformer=transformer,
                    torch_dtype=torch.bfloat16,
                    )
                need_clip=True
            else:
                pipeline = FluxPipeline_orgin.from_single_file(
                    flux_transformer_path,
                    config=flux_repo_local,
                    torch_dtype=torch.bfloat16,
                    )
            
        else:
            # flux_repo='F:/test/ComfyUI/models/diffusers/black-forest-labs/FLUX.1-dev'
            if quantization=="none":
                pipeline =FluxPipeline_orgin.from_pretrained(flux_repo,torch_dtype=torch.bfloat16,)
            else:
                if quantization=="fp8":
                    transformer = FluxTransformer2DModel.from_pretrained(
                        flux_repo,
                        subfolder="transformer",
                        quantization_config=DiffusersBitsAndBytesConfig(load_in_8bit=True,),
                        torch_dtype=torch.bfloat16,
                    )
                else: #nf4
                    transformer = FluxTransformer2DModel.from_pretrained(
                        flux_repo,
                        subfolder="transformer",
                        quantization_config=DiffusersBitsAndBytesConfig(
                            load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=torch.bfloat16
                        ),
                        torch_dtype=torch.bfloat16,)
                pipeline =FluxPipeline_orgin.from_pretrained(flux_repo,transformer=transformer,torch_dtype=torch.bfloat16,)

        # Load and fuse base LoRA weights
        print("***********  Load and fuse base LoRA weights ***********")
        pipeline.load_lora_weights(lora_path)
        pipeline.fuse_lora()
        pipeline.unload_lora_weights()
        pipeline.load_lora_weights(pre_lora_path)
        pipeline.enable_model_cpu_offload()

        print("***********Load model done ***********")

        cleanup()

        if use_mmgp:
            from mmgp import offload as offloadobj
            pipe = {"transformer": pipeline, }
            # offloadobj.profile(pipe, quantizeTransformer = False,  profile_no = 1 ) # uncomment this line and comment the previous one if you have 24 GB of VRAM and wants faster generation  
            offloadobj.profile(pipe, quantizeTransformer = False,  extraModelsToQuantize = [], profile_no = profile_number, ) 
            del pipe
        return ({"pipeline":pipeline,"need_clip":need_clip},)



class PhotoDoodle_Sampler:
    def __init__(self):
        pass

    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "model": ("MODEL_PhotoDoodle",),
                "images": ("IMAGE",),  
                "prompt": ("STRING", {"default": "add a halo and wings for the cat by sksmagiceffects", "multiline": True}),
                "width": ("INT", {"default": 512, "min": 256, "max": 4096, "step": 64, "display": "number"}),
                "height": ("INT", {"default": 768, "min": 256, "max": 4096, "step": 64, "display": "number"}),
                "steps": ("INT", {"default": 20, "min": 1, "max": 1024, "step": 1, "display": "number"}),
                "guidance_scale": ("FLOAT", {"default": 3.5, "min": 0.0, "max": 10.0, "step": 0.1}),
                "max_sequence_length": ("INT", {"default": 512, "min": 128, "max": 512, "step": 1, "display": "number"}),
                "seed": ("INT", {"default": -1, "min": -1, "max": 4294967295}),
                },
            "optional": { "clip":("CLIP",),},
            }

    RETURN_TYPES = ("IMAGE", )
    RETURN_NAMES = ("image",)
    FUNCTION = "sampler_main"
    CATEGORY = "PhotoDoodle"

    def sampler_main(self, model,images,prompt, width, height,steps,guidance_scale, max_sequence_length,seed=-1,**kwargs):
        # 设置随机种子
        if seed == -1:
            seed = torch.randint(0, 4294967295, (1,)).item()
        generator = torch.Generator(device=device).manual_seed(seed)
        
        need_clip=model.get("need_clip")
        pipeline=model.get("pipeline")
        clip=kwargs.get("clip")
        # load model
        #model.to(device)
        if need_clip:
            if clip is None:
                raise ValueError("No clip selected")
            # fix pooled of comfyclip
            tokens = clip.tokenize(prompt)
            tokens["t5xxl"] = clip.tokenize(prompt)["t5xxl"]
            prompt_embeds = clip.encode_from_tokens(tokens, return_pooled=True, return_dict=True).pop("cond")
            tokens["l"] = clip.tokenize(prompt)["l"]
            pooled_prompt_embeds = clip.encode_from_tokens(tokens,  return_dict=True).pop("pooled_output")
            #prompt_embeds,pooled_prompt_embeds=clip.encode_from_tokens( clip.tokenize(prompt,return_word_ids=True),return_pooled=True)
            #print(prompt_embeds.shape,pooled_prompt_embeds.shape)
            prompt_embeds=prompt_embeds.to(device,torch.bfloat16)
            pooled_prompt_embeds=pooled_prompt_embeds.to(device,torch.bfloat16)
            prompt=None
            del clip

        else:
            prompt_embeds,pooled_prompt_embeds=None,None

        cleanup()
        condition_image_list = tensor2pil_list(images, width, height)
        total_images=len(condition_image_list)
        pbar = ProgressBar(total_images)
        img_list=[]

        for i, condition_image in enumerate(condition_image_list):
            result = pipeline(
                prompt=prompt,
                prompt_embeds=prompt_embeds,
                pooled_prompt_embeds=pooled_prompt_embeds,
                condition_image=condition_image,
                height=height,
                width=width,
                guidance_scale=guidance_scale,
                num_inference_steps=steps,
                max_sequence_length=max_sequence_length,
                generator=generator,  # 添加 generator 参数
            ).images[0]
            pbar.update_absolute(i, total_images)
            img_list.append(result)
            
        cleanup()
        return (load_images(img_list),)


NODE_CLASS_MAPPINGS = {
    "PhotoDoodle_Loader": PhotoDoodle_Loader,
    "PhotoDoodle_Sampler": PhotoDoodle_Sampler,
}

NODE_DISPLAY_NAME_MAPPINGS = {
    "PhotoDoodle_Loader": "PhotoDoodle_Loader",
    "PhotoDoodle_Sampler": "PhotoDoodle_Sampler",
}
]

**Comment by smthemex** [comment]
TKS，I will add it
